scripts/engine-dag/main.go

```diff
@@ -0,0 +1,191 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package main
+
+import (
+	"bytes"
+	"flag"
+	"fmt"
+	"strings"
+
+	"github.com/juju/collections/set"
+	"github.com/juju/names/v5"
+	"github.com/juju/worker/v4/dependency"
+
+	"github.com/juju/juju/agent"
+	"github.com/juju/juju/cmd/jujud-controller/agent/machine"
+	"github.com/juju/juju/controller"
+	"github.com/juju/juju/internal/testing"
+	"github.com/juju/juju/internal/upgrades"
+	"github.com/juju/juju/state"
+)
+
+var (
+	modelTypeFlag      = flag.String("model-type", "iaas", "model type to use (iaas|caas)")
+	transitiveDepsFlag = flag.Int("transitive-deps", 0, "include transitive dependencies and how many levels to include")
+	manifoldFlag       = flag.String("manifold", "", "manifold to select (empty indicates all)")
+)
+
+func main() {
+	flag.Parse()
+
+	var manifolds dependency.Manifolds
+	switch *modelTypeFlag {
+	case "iaas":
+
+		manifolds = machine.IAASManifolds(machine.ManifoldsConfig{
+			Agent:           &mockAgent{},
+			PreUpgradeSteps: preUpgradeSteps,
+		})
+	case "caas":
+		manifolds = machine.CAASManifolds(machine.ManifoldsConfig{
+			Agent:           &mockAgent{},
+			PreUpgradeSteps: preUpgradeSteps,
+		})
+	default:
+		panic("unknown model type")
+	}
+
+	root := NewDAG()
+
+	transativeDepth := *transitiveDepsFlag
+	selectedManifold := *manifoldFlag
+	for name, manifold := range manifolds {
+		if selectedManifold != "" && name != selectedManifold {
+			continue
+		}
+		node := root.AddVertex(name)
+		dependencies := manifoldDependencies(manifolds, manifold, transativeDepth)
+		for _, dep := range dependencies.Values() {
+			node.AddEdge(dep)
+		}
+	}
+
+	fmt.Println(root.Render())
+}
+
+func manifoldDependencies(all dependency.Manifolds, manifold dependency.Manifold, transativeDepth int) set.Strings {
+	result := set.NewStrings()
+	for _, input := range manifold.Inputs {
+		result.Add(input)
+		if transativeDepth == 0 {
+			continue
+		}
+
+		result = result.Union(manifoldDependencies(all, all[input], transativeDepth-1))
+	}
+	return result
+}
+
+type Dag struct {
+	nodes map[string]*DagNode
+}
+
+func NewDAG() *Dag {
+	root := new(Dag)
+	root.nodes = make(map[string]*DagNode)
+	return root
```

Comment:
**nitpick:** I prefer to not use `new` which seems so weird in golang.

```go
```suggestion

func NewDAG() *Dag {
    return &Dag{
        nodes: make(map[string]*DagNode),
    }
}
```